internal/auth/providers/okta.go

```diff
@@ -0,0 +1,423 @@
+package providers
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/buzzfeed/sso/internal/auth/circuit"
+	log "github.com/buzzfeed/sso/internal/pkg/logging"
+	"github.com/buzzfeed/sso/internal/pkg/sessions"
+	"github.com/datadog/datadog-go/statsd"
+)
+
+// OktaProvider is an implementation of the Provider interface.
+type OktaProvider struct {
+	*ProviderData
+	StatsdClient *statsd.Client
+	cb           *circuit.Breaker
+}
+
+type GetUserProfileResponse struct {
+	EmailAddress  string   `json:"email"`
+	EmailVerified bool     `json:"email_verified"`
+	Groups        []string `json:"groups"`
+}
+
+// NewOktaProvider returns a new OktaProvider and sets the provider url endpoints.
+func NewOktaProvider(p *ProviderData, orgName, providerServerID string) (*OktaProvider, error) {
+	if orgName != "" || providerServerID != "" {
+		if orgName == "" {
+			return nil, errors.New("missing setting: org_name")
+		}
+		if providerServerID == "" {
+			return nil, errors.New("missing setting: provider_server_id")
+		}
+	}
+
+	p.ProviderName = "Okta"
+	// interact with resource owner and obtain an authorization grant
+	if p.SignInURL.String() == "" {
+		p.SignInURL = &url.URL{
+			Scheme: "https",
+			Host:   fmt.Sprintf("%s.okta.com", strings.Trim(orgName, "\"")),
```

Comment:
Let's move this line to a higher scope rather than repeating on every url construction block, same with `providerServerID` and `Scheme`.

```golang
scheme := "https"
host := fmt.Sprintf("%s.okta.com", strings.Trim(orgName, "\"")),
providerServerID := strings.Trim(providerServerID, "\"")
. . .
if p.SignInURL.String() == "" {
    p.SignInURL = &url.URL{
        Scheme: scheme,
        Host:   host,
        Path:   fmt.Sprintf("/oauth2/%s/v1/authorize", providerServerID),
    }
}
```

Secondly, why do we need to trim `"` from all these strings? It's a little weird to be doing at the provider module level. We should be stripping any lingering `"` at the options parse layer.

Reply:
👍 I also moved the string trimming to the options parse layer.